tf2_geometry_msgs geometry_msgs to TF2 StampedTransform

<!--
For general questions, please ask on ROS answers: https://answers.ros.org, make sure to include at least the `ros2` tag and the rosdistro version you are running, e.g. `ardent`.
For general design discussions, please post on discourse: https://discourse.ros.org/c/ng-ros
Not sure if this is the right repository? Open an issue on https://github.com/ros2/ros2/issues
For Bug report or feature requests, please fill out the relevant category below
-->

## Bug report

**Required Info:**

- Operating System: Ubuntu 18.04
  - <!-- OS and version (e.g. Windows 10, Ubuntu 16.04...) -->
- Installation type: Dashing latest release, E master on hash below
  - <!-- binaries or from source  -->
- Version or commit hash: 4e519f3e77611a18f03398b6334e9cc8ccf49552
  - <!-- Output of git rev-parse HEAD, release version, or repos file  -->
- DDS implementation: Fast RTPS
  - <!-- rmw_implementation used (e.g. Fast-RTPS, RTI Connext, etc -->
- Client library (if applicable): rclcpp
  - <!-- e.g. rclcpp, rclpy, or N/A -->

#### Steps to reproduce issue
<!-- Detailed instructions on how to reliably reproduce this issue http://sscce.org/
``` code that can be copy-pasted is preferred ``` -->

try to call

```
geometry_msgs::msg::TransformStamped toMsg(const tf2::Stamped<tf2::Transform>& in)
```
Defined here: https://github.com/ros2/geometry2/blob/ros2/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h#L513

#### Expected behavior

I'm trying to convert a bunch of types around, and I've noticed a weird issue where all my other conversions work fine, except for this one. I get errors stating that this particular overload of `toMsg` doesnt exist. It works when I get out only a Transform, but not a stamped transform. I attempted both on Dashing from debians and on my ros2 master build, either recognize this and given it was written awhile ago I'm a bit confused why its not working.

#### Actual behavior
Typical compiler error saying it can't be found.  

```
/home/s.macenski/Documents/slam_ws/src/slam_toolbox/src/slam_toolbox_common.cpp: In member function ‘tf2::Stamped<tf2::Transform> slam_toolbox::SlamToolbox::setTransformFromPoses(const karto::Pose2&, const karto::Pose2&, const rclcpp::Time&, const bool&)’:
/home/s.macenski/Documents/slam_ws/src/slam_toolbox/src/slam_toolbox_common.cpp:398:45: error: no match for ‘operator=’ (operand types are ‘geometry_msgs::msg::TransformStamped {aka geometry_msgs::msg::TransformStamped_<std::allocator<void> >}’ and ‘geometry_msgs::msg::Transform {aka geometry_msgs::msg::Transform_<std::allocator<void> >}’)
     base_to_map_msg = tf2::toMsg(base_to_map);
```

#### Additional information


Specific set of code, while migrating Slam Toolbox to ROS2. Its ROS1 counterpart using `tf2::convert` while uses these methods under the hood has no issues. 
```
  // Compute the map->odom transform
  tf2::Stamped<tf2::Transform> odom_to_map;
  tf2::Quaternion q(0.,0.,0.,1.0);
  q.setRPY(0., 0., corrected_pose.GetHeading());
  tf2::Stamped<tf2::Transform> base_to_map(
    tf2::Transform(q, tf2::Vector3(corrected_pose.GetX(),
    corrected_pose.GetY(), 0.0)).inverse(), tf2_ros::fromMsg(t), base_frame_);
  try
  {
    geometry_msgs::msg::TransformStamped base_to_map_msg, odom_to_map_msg;
    base_to_map_msg = tf2::toMsg(base_to_map); // this fails

    // this stuff works, obviously
    // base_to_map_msg.header.stamp = tf2_ros::toMsg(base_to_map.stamp_);
    // base_to_map_msg.header.frame_id = base_to_map.frame_id_;
    // base_to_map_msg.transform.translation.x = base_to_map.getOrigin().getX();
    // base_to_map_msg.transform.translation.y = base_to_map.getOrigin().getY();
    // base_to_map_msg.transform.translation.z = base_to_map.getOrigin().getZ();
    // base_to_map_msg.transform.rotation = tf2::toMsg(base_to_map.getRotation());

    odom_to_map_msg = tf_->transform(base_to_map_msg, odom_frame_);
    tf2::fromMsg(odom_to_map_msg, odom_to_map);
  }
  catch(tf2::TransformException & e)
  {
    RCLCPP_ERROR(get_logger(), "Transform from base_link to odom failed: %s",
      e.what());
    odom_to_map.setIdentity();
  }
```

And its counterpart, fromMsg, works fine between `tf2::stamped<tf2::transform>` and `geometry_msgs/transformstamped`

--- a/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h
+++ b/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h
@@ -354,7 +354,6 @@ void fromMsg(const geometry_msgs::msg::QuaternionStamped& msg, geometry_msgs::ms
  * \param in A instance of the tf2::Quaternion specialization of the tf2::Stamped template.
  * \return The QuaternionStamped converted to a geometry_msgs QuaternionStamped message type.
  */
-template <>
 inline
 geometry_msgs::msg::QuaternionStamped toMsg(const tf2::Stamped<tf2::Quaternion>& in)
 {

--- a/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h
+++ b/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h
@@ -372,7 +372,6 @@ geometry_msgs::msg::QuaternionStamped toMsg(const tf2::Stamped<tf2::Quaternion>&
  * \param in A QuaternionStamped message type.
  * \param out The QuaternionStamped converted to the equivalent tf2 type.
  */
-template <>
 inline
 void fromMsg(const geometry_msgs::msg::QuaternionStamped& in, tf2::Stamped<tf2::Quaternion>& out)
 {

--- a/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h
+++ b/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h
@@ -508,7 +508,6 @@ void fromMsg(const geometry_msgs::msg::TransformStamped& in, tf2::Stamped <tf2::
  * \param in An instance of the tf2::Transform specialization of the tf2::Stamped template.
  * \return The TransformStamped converted to a geometry_msgs TransformStamped message type.
  */
-template <>
 inline
 geometry_msgs::msg::TransformStamped toMsg(const tf2::Stamped<tf2::Transform>& in)
 {

--- a/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
+++ b/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
@@ -40,6 +40,21 @@ static const double EPS = 1e-3;
 
 TEST(TfGeometry, Conversions)
 {
+  // QuaternionStamped
+  {
+    auto rotation = tf2::Quaternion(1.0, 2.0, 3.0, 4.0).normalized();
+    auto stamp = tf2::timeFromSec(2);
+    std::string frame_id = "test_frame_id";
+    tf2::Stamped<tf2::Quaternion> quat_stamped(rotation, stamp, frame_id);
+    geometry_msgs::msg::QuaternionStamped quat_stamped_msg;
+    quat_stamped_msg = tf2::toMsg(quat_stamped);
+
+    EXPECT_NEAR(rotation.getX(), quat_stamped_msg.quaternion.x, EPS);
+    EXPECT_NEAR(rotation.getY(), quat_stamped_msg.quaternion.y, EPS);
+    EXPECT_NEAR(rotation.getZ(), quat_stamped_msg.quaternion.z, EPS);
+    EXPECT_NEAR(rotation.getW(), quat_stamped_msg.quaternion.w, EPS);
+  }
+
   // TransformStamped
   {
     auto rotation = tf2::Quaternion(1.0, 2.0, 3.0, 4.0).normalized();

--- a/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
+++ b/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
@@ -47,7 +47,7 @@ TEST(TfGeometry, Conversions)
     std::string frame_id = "test_frame_id";
     tf2::Stamped<tf2::Quaternion> quat_stamped(rotation, stamp, frame_id);
     geometry_msgs::msg::QuaternionStamped quat_stamped_msg;
-    quat_stamped_msg = tf2::toMsg(quat_stamped);
+    tf2::convert(quat_stamped, quat_stamped_msg);
 
     EXPECT_NEAR(rotation.getX(), quat_stamped_msg.quaternion.x, EPS);
     EXPECT_NEAR(rotation.getY(), quat_stamped_msg.quaternion.y, EPS);
@@ -56,7 +56,7 @@ TEST(TfGeometry, Conversions)
     EXPECT_EQ(frame_id, quat_stamped_msg.header.frame_id);
 
     tf2::Stamped<tf2::Quaternion> quat_from_msg;
-    tf2::fromMsg(quat_stamped_msg, quat_from_msg);
+    tf2::convert(quat_stamped_msg, quat_from_msg);
 
     EXPECT_NEAR(quat_from_msg.getX(), quat_stamped_msg.quaternion.x, EPS);
     EXPECT_NEAR(quat_from_msg.getY(), quat_stamped_msg.quaternion.y, EPS);
@@ -73,7 +73,7 @@ TEST(TfGeometry, Conversions)
     std::string frame_id = "test_frame_id";
     tf2::Stamped<tf2::Transform> tf_stamped(tf2::Transform(rotation, translation), stamp, frame_id);
     geometry_msgs::msg::TransformStamped tf_stamped_msg;
-    tf_stamped_msg = tf2::toMsg(tf_stamped);
+    tf2::convert(tf_stamped, tf_stamped_msg);
 
     EXPECT_NEAR(rotation.getX(), tf_stamped_msg.transform.rotation.x, EPS);
     EXPECT_NEAR(rotation.getY(), tf_stamped_msg.transform.rotation.y, EPS);
@@ -85,7 +85,7 @@ TEST(TfGeometry, Conversions)
     EXPECT_EQ(frame_id, tf_stamped_msg.header.frame_id);
 
     tf2::Stamped<tf2::Transform> tf_from_msg;
-    tf2::fromMsg(tf_stamped_msg, tf_from_msg);
+    tf2::convert(tf_stamped_msg, tf_from_msg);
 
     EXPECT_NEAR(tf_from_msg.getRotation().getX(), tf_stamped_msg.transform.rotation.x, EPS);
     EXPECT_NEAR(tf_from_msg.getRotation().getY(), tf_stamped_msg.transform.rotation.y, EPS);

--- a/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
+++ b/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
@@ -42,7 +42,7 @@ TEST(TfGeometry, Conversions)
 {
   // TransformStamped
   {
-    tf2::Quaternion rotation(1.0, 2.0, 3.0, 4.0);
+    auto rotation = tf2::Quaternion(1.0, 2.0, 3.0, 4.0).normalized();
     tf2::Vector3 translation(1.0, 2.0, 3.0);
     auto stamp = tf2::timeFromSec(2);
     std::string frame_id = "test_frame_id";
@@ -50,13 +50,13 @@ TEST(TfGeometry, Conversions)
     geometry_msgs::msg::TransformStamped tf_stamped_msg;
     tf_stamped_msg = tf2::toMsg(tf_stamped);
 
-    EXPECT_NEAR(rotation.x(), tf_stamped_msg.transform.rotation.x, EPS);
-    EXPECT_NEAR(rotation.y(), tf_stamped_msg.transform.rotation.y, EPS);
-    EXPECT_NEAR(rotation.z(), tf_stamped_msg.transform.rotation.z, EPS);
-    EXPECT_NEAR(rotation.w(), tf_stamped_msg.transform.rotation.w, EPS);
-    EXPECT_NEAR(translation.x(), tf_stamped_msg.transform.translation.z, EPS);
-    EXPECT_NEAR(translation.y(), tf_stamped_msg.transform.translation.y, EPS);
-    EXPECT_NEAR(translation.z(), tf_stamped_msg.transform.translation.z, EPS);
+    EXPECT_NEAR(rotation.getX(), tf_stamped_msg.transform.rotation.x, EPS);
+    EXPECT_NEAR(rotation.getY(), tf_stamped_msg.transform.rotation.y, EPS);
+    EXPECT_NEAR(rotation.getZ(), tf_stamped_msg.transform.rotation.z, EPS);
+    EXPECT_NEAR(rotation.getW(), tf_stamped_msg.transform.rotation.w, EPS);
+    EXPECT_NEAR(translation.getX(), tf_stamped_msg.transform.translation.x, EPS);
+    EXPECT_NEAR(translation.getY(), tf_stamped_msg.transform.translation.y, EPS);
+    EXPECT_NEAR(translation.getZ(), tf_stamped_msg.transform.translation.z, EPS);
   }
 }
 

--- a/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
+++ b/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
@@ -38,6 +38,27 @@
 std::unique_ptr<tf2_ros::Buffer> tf_buffer = nullptr;
 static const double EPS = 1e-3;
 
+TEST(TfGeometry, Conversions)
+{
+  // TransformStamped
+  {
+    tf2::Quaternion rotation(1.0, 2.0, 3.0, 4.0);
+    tf2::Vector3 translation(1.0, 2.0, 3.0);
+    auto stamp = tf2::timeFromSec(2);
+    std::string frame_id = "test_frame_id";
+    tf2::Stamped<tf2::Transform> tf_stamped(tf2::Transform(rotation, translation), stamp, frame_id);
+    geometry_msgs::msg::TransformStamped tf_stamped_msg;
+    tf_stamped_msg = tf2::toMsg(tf_stamped);
+
+    EXPECT_NEAR(rotation.x(), tf_stamped_msg.transform.rotation.x, EPS);
+    EXPECT_NEAR(rotation.y(), tf_stamped_msg.transform.rotation.y, EPS);
+    EXPECT_NEAR(rotation.z(), tf_stamped_msg.transform.rotation.z, EPS);
+    EXPECT_NEAR(rotation.w(), tf_stamped_msg.transform.rotation.w, EPS);
+    EXPECT_NEAR(translation.x(), tf_stamped_msg.transform.translation.z, EPS);
+    EXPECT_NEAR(translation.y(), tf_stamped_msg.transform.translation.y, EPS);
+    EXPECT_NEAR(translation.z(), tf_stamped_msg.transform.translation.z, EPS);
+  }
+}
 
 TEST(TfGeometry, Frame)
 {

--- a/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
+++ b/tf2_geometry_msgs/test/test_tf2_geometry_msgs.cpp
@@ -53,6 +53,16 @@ TEST(TfGeometry, Conversions)
     EXPECT_NEAR(rotation.getY(), quat_stamped_msg.quaternion.y, EPS);
     EXPECT_NEAR(rotation.getZ(), quat_stamped_msg.quaternion.z, EPS);
     EXPECT_NEAR(rotation.getW(), quat_stamped_msg.quaternion.w, EPS);
+    EXPECT_EQ(frame_id, quat_stamped_msg.header.frame_id);
+
+    tf2::Stamped<tf2::Quaternion> quat_from_msg;
+    tf2::fromMsg(quat_stamped_msg, quat_from_msg);
+
+    EXPECT_NEAR(quat_from_msg.getX(), quat_stamped_msg.quaternion.x, EPS);
+    EXPECT_NEAR(quat_from_msg.getY(), quat_stamped_msg.quaternion.y, EPS);
+    EXPECT_NEAR(quat_from_msg.getZ(), quat_stamped_msg.quaternion.z, EPS);
+    EXPECT_NEAR(quat_from_msg.getW(), quat_stamped_msg.quaternion.w, EPS);
+    EXPECT_EQ(quat_from_msg.frame_id_, quat_stamped_msg.header.frame_id);
   }
 
   // TransformStamped
@@ -72,6 +82,19 @@ TEST(TfGeometry, Conversions)
     EXPECT_NEAR(translation.getX(), tf_stamped_msg.transform.translation.x, EPS);
     EXPECT_NEAR(translation.getY(), tf_stamped_msg.transform.translation.y, EPS);
     EXPECT_NEAR(translation.getZ(), tf_stamped_msg.transform.translation.z, EPS);
+    EXPECT_EQ(frame_id, tf_stamped_msg.header.frame_id);
+
+    tf2::Stamped<tf2::Transform> tf_from_msg;
+    tf2::fromMsg(tf_stamped_msg, tf_from_msg);
+
+    EXPECT_NEAR(tf_from_msg.getRotation().getX(), tf_stamped_msg.transform.rotation.x, EPS);
+    EXPECT_NEAR(tf_from_msg.getRotation().getY(), tf_stamped_msg.transform.rotation.y, EPS);
+    EXPECT_NEAR(tf_from_msg.getRotation().getZ(), tf_stamped_msg.transform.rotation.z, EPS);
+    EXPECT_NEAR(tf_from_msg.getRotation().getW(), tf_stamped_msg.transform.rotation.w, EPS);
+    EXPECT_NEAR(tf_from_msg.getOrigin().getX(), tf_stamped_msg.transform.translation.x, EPS);
+    EXPECT_NEAR(tf_from_msg.getOrigin().getY(), tf_stamped_msg.transform.translation.y, EPS);
+    EXPECT_NEAR(tf_from_msg.getOrigin().getZ(), tf_stamped_msg.transform.translation.z, EPS);
+    EXPECT_EQ(tf_from_msg.frame_id_, tf_stamped_msg.header.frame_id);
   }
 }
 

